src/main/solution_space/path_finder/path_finder_v_3.py

```diff
@@ -41,63 +51,58 @@ def find_next_anon_tree(self, user_anon_tree: AnonTree, user_canon_tree: ast.AST
-        """
-        goals: List[Vertex] = self._graph.end_vertex.parents
-        goals_anon_trees = sum([g.serialized_code.anon_trees for g in goals], [])
-
-        log.info(f'Number of goals: {len(goals)}\nGoals ids are {[g.id for g in goals]}')
-        return self.__choose_best_anon_tree(user_anon_tree, goals_anon_trees)
-
-    # Note: we have to remove the 'user_code' from the set
-    def __find_closest_tree_in_graph(self, user_anon_tree: AnonTree, user_canon_tree: ast.AST) -> Optional[AnonTree]:
-        """
-        1. If there is a vertex in the graph with same canon_tree:
-            1.1 *find out what to do*
-        2. Consider each vertex with similar nodes number as candidate (chose at least TOP_N_CANON candidates)
-        """
-
         graph_vertex = self._graph.find_vertex(user_canon_tree)
-        # Todo: find a better way
         if graph_vertex:
             graph_anon_tree = graph_vertex.serialized_code.find_anon_tree(user_anon_tree.tree)
             if graph_anon_tree:
-                return self.__find_closest_tree_from_equal_tree(user_anon_tree, graph_anon_tree)
+                next_anon_trees = [AnonTree.get_item_by_id(id) for id in graph_anon_tree.next_anon_trees_ids]
+                return self.__choose_best_anon_tree(user_anon_tree, next_anon_trees)
+        return None
 
-        canon_nodes_number = get_nodes_number_in_ast(user_canon_tree)
-        anon_nodes_number = user_anon_tree.nodes_number
+    @staticmethod
+    def __get_nodes_number_dict(nodes_number: List[Any], filter_foo: Callable = lambda x: x) -> Dict[int, list]:
+        return {k: list(v) for k, v in groupby(nodes_number, filter_foo)}
+
+    # Note: we have to remove the 'user_code' from the set
+    def __find_closest_tree(self, user_anon_tree: AnonTree, user_canon_nodes_number: int,
+                            user_anon_tree_nodes_number: int,
+                            canon_trees_nodes_number: Dict[int, list]) -> Optional[AnonTree]:
+        """
+        1. Consider each vertex with similar nodes number as candidate (chose at least TOP_N_CANON candidates)
+        """
 
         # Get vertices ids with canon trees, which have nodes number similar to user canon_nodes_number
-        vertices_ids = self.__get_top_n_candidates(CANON_TOP_N, canon_nodes_number, self._graph.canon_trees_nodes_number)
+        vertices_ids = self.__get_top_n_candidates(CANON_TOP_N, user_canon_nodes_number, canon_trees_nodes_number)
         vertices: List[Vertex] = [Vertex.get_item_by_id(id) for id in vertices_ids]
 
         anon_trees = sum([v.serialized_code.anon_trees for v in vertices], [])
-        anon_nodes_numbers_dict = {k: list(v) for k, v in groupby(anon_trees, lambda a_t: a_t.nodes_number)}
-        anon_candidates = self.__get_top_n_candidates(ANON_TOP_N, anon_nodes_number, anon_nodes_numbers_dict)
-
+        anon_nodes_numbers_dict = self.__class__.__get_nodes_number_dict(anon_trees, lambda a_t: a_t.nodes_number)
```

Comment:
You can call classmethods without calling `__class__` because `self` is an instance of class. So just call self.__get_nodes_number_dict